### ISSUE TITLE
USB Wifi Hotspot

### ISSUE THREAD

**Issue by bousqi** [Mwyann/psakey#1]
Hi Mwyann, 

I'm reaching you as I found your work on PSAKEY.
Have you ever tried to make a simple Wifi Hotspot with your CDC_EEM Usb Key (based on a RPI Zero) ?
Such hotspot would allow to easily connect a computer to SMEG+ telnet shell.

I'm currently playing a little bit with my SMEG+ unit in my car, trying to understand how it works. 
You can find my actual work in my [SMEG_PLUS](https://github.com/bousqi/SMEG_PLUS) repo.
My work is based on a Bluetooth link and replacing it by a wifi link would be better. I won't have to sit in the car to test few things.

I'm even thinking of a RPI Zero W that would feed last ZAR over wifi and replace EEM Gadget by a Mass Storage Gadget to automatically install it when available.

**Comment by bousqi** [comment]
Interesting.
On my side I'm making a readable disassembly of vxWorks. 
There are far more commands than the ones listed in help. Once my vxWorks analysis will be complete, I'll publish a complete list. 

Regarding NFS mount, I guess you mounted RPI remote storage from vxWorks ?
Have you tried the opposite ? 

I also used nmap to check open ports, but I stayed in a low range.
Do you have any idea about 20000 port usage ? with a netstat command (ipnet_cmd_netstat) on SMEG+ I'have seen that 20000 port range seems to be used but locally.

Last but not least, can you provide more details on the RPI0W config ?
Have you been using a specific configuration for CDC_EEM to be recognized by SMEG+ ? 
Any filtering on VID/PID ? I have a long list of USB EEM device names in vxWorks, so I guess it can't be a random one.

**Comment by Mwyann** [comment]
Great, looking forward that new list :)

About NFS, that's right, I connected vxWorks into the RPI, did not try the opposite (I'm not even sure it's possible).

I don't know what port 20000 is for.

The RPI0W USB gadget config is located in the  myusbgadget file here: https://github.com/Mwyann/psakey/blob/master/resources/myusbgadget
Those are taken from the PSA Key, maybe somoe other VID/PID will be recognized and maybe they may expose additionnal services, why not (like if some debugging tools could be plugged into the USB port).

Yann

**Comment by bousqi** [comment]
Thanks for point out these details. 
I'll try this when my RPI0W will be recevied !

By any chance, do you have some spare memories (or a log) of commands to mount an NFS from vxWorks to RPI ?
That could be interesting to write it down. Some people might find a usage for that.

**Comment by bousqi** [comment]
I recieved my ZeroW and i'm following your guide to setup the key.
I understand that usb_f_eem is the only module not prebuilt in raspbian ? So we need to build it ourselves.
Have you tried to rely on other USB gadget like RNDIS/NCM/ECM (rather than EEM) ?

I just had a look to vxWorks disassembly and I can only find references of EEM support.

One question, can you argue on the need to set the filesystem in RO ? a comment indicates that the Rpi might reboot without notice. Why ?

**Comment by bousqi** [comment]
Thanks for you feedback, I'll save a lot of time not testing other Ethernet models over USB.
I guess EEM gadget source might not change so much, I wonder if it won't be possible to make a DKMS to automatically build the usb_f_eem.ko.

I'll build the module for the current kernel, and push it to your git, and try to play with DKMS.

**Comment by bousqi** [comment]
Funny thing, not so funny.
I realized that Mass Storage gadget is not working if the VID/PID of the PSAKey is used.
vxWorks must associate this usb token ids to EEM only and refuses to mount MassStorage.

**Comment by Mwyann** [comment]
You're right :) You cannot use both Mass Storage and EEM (even in composite mode it doesn't work). So if you want to switch between the two, you have to reset the USB gadget config.

**Comment by bousqi** [comment]
I did try to use Mass Storage only with this config : 

```
echo 0x243a > idVendor  # Peugeot
echo 0x0001 > idProduct # Connect Apps
echo 0x0226 > bcdDevice # v2.26
echo 0x0200 > bcdUSB    # USB 2.0

echo "CAFEDECA"                  > strings/0x409/serialnumber
echo "Aperture Science"          > strings/0x409/manufacturer
echo "PSAKey"                    > strings/0x409/product
```

But SMEG+ refuses to scan it for storage.
Thus I had to duplicate a USB Key VID/PID and strings

**Comment by vidarrt9** [comment]
> I realized that Mass Storage gadget is not working if the VID/PID of the PSAKey is used.
> vxWorks must associate this usb token ids to EEM only and refuses to mount MassStorage.

This may work if you use [`dummy_hcd`](https://www.collabora.com/news-and-blog/blog/2019/06/24/using-dummy-hcd/). Basically when you pose as CDC_EEM you offer only a single interface (in USB terminology). If you use `dummy_hcd` you pose as a USB hub which can then in turn provide a pathway to additional interfaces.

You may want to investigate [the Facedancer project](https://github.com/usb-tools/Facedancer) also.

**Comment by Mwyann** [comment]
> > I realized that Mass Storage gadget is not working if the VID/PID of the PSAKey is used.
> > vxWorks must associate this usb token ids to EEM only and refuses to mount MassStorage.
> 
> This may work if you use [`dummy_hcd`](https://www.collabora.com/news-and-blog/blog/2019/06/24/using-dummy-hcd/). Basically when you pose as CDC_EEM you offer only a single interface (in USB terminology). If you use `dummy_hcd` you pose as a USB hub which can then in turn provide a pathway to additional interfaces.
> 
> You may want to investigate [the Facedancer project](https://github.com/usb-tools/Facedancer) also.

If I understand correctly, dummy_hcd isn't a USB function for gadgets, but it allows to set up a fake USB host so you can try USB gadgets locally without having compatible hardware (i.e. a proper USB OTG interface). This is different from a fake USB hub gadget function that you can use to attach more than one (virtual) device to it. And I couldn't find that too.

Also, when the original USB key was still working, I tried using a real USB hub with the PSA key and some flash stick, to see if I could use both, but it didn't work, even when powered externally (from the lighter socket). Maybe my hub was faulty, or the power supply I was using was not good enough (the PSA key needed some proper voltages to work), but I concluded that the car doesn't react if the "root device" is not what it's waiting for.

**Comment by bousqi** [comment]
SMEG Unit USB controller is very sensitive. You should not try to draw power from it. 
My RPI zerow make the units reboot.

I'll definitely give a try to dummy_hcd. That sounds promising. I'm always pleased to realized how far we can go in linux kernel to simulate USB devices. Developping USB devices could be such a nightmare ! And when I did it, it was only about USB2 :)

**Comment by vidarrt9** [comment]
> If I understand correctly, dummy_hcd isn't a USB function for gadgets, but it allows to set up a fake USB host so you can try USB gadgets locally without having compatible hardware (i.e. a proper USB OTG interface). This is different from a fake USB hub gadget function that you can use to attach more than one (virtual) device to it. And I couldn't find that too.

I see. Perhaps I misinterpreted it. I'll see it when I get to try.

> Also, when the original USB key was still working, I tried using a real USB hub with the PSA key and some flash stick, to see if I could use both, but it didn't work, even when powered externally (from the lighter socket). Maybe my hub was faulty, or the power supply I was using was not good enough (the PSA key needed some proper voltages to work), but I concluded that the car doesn't react if the "root device" is not what it's waiting for.

That's well possible. Which exact hardware version do you have? As far as I read only the version with two USB ports may be able to do _that_.

@bousqi Facedancer with the appropriate hardware is your best bet to get started quickly.

**Comment by Mwyann** [comment]
> SMEG Unit USB controller is very sensitive. You should not try to draw power from it.
> My RPI zerow make the units reboot.

Really? Mine has only one USB port, but the PSA key advertised a MaxPower of 500 mA on the USB bus, and I used a RPI0w and also an OrangePi (which I currently use) without any problem at all. My OrangePi even have an extra USB Wifi dongle (because the integrated Wifi controller is garbage) and a USB BT dongle for the OBD2 communication ;-)

**Comment by Mwyann** [comment]
> That's well possible. Which exact hardware version do you have? As far as I read only the version with two USB ports may be able to do _that_.

I don't know exactly which version I have. The upgrade screen tells me my hardware ID is 155. And I have only one USB port.
